<h1 style="color: #001a79;">Topic 4 - Machine Learning Project Notebook</h1>

In [4]:
# Efficient numerical arrays.
import numpy as np

# Dataframes for data manipulation and analysis.
import pandas as pd

# Neural networks.
import tensorflow.keras as keras

# Plotting.
import matplotlib.pyplot as plt

<h3 style="color: #001a79;">Introduction</h3>

<hr style="border-top: 1px solid #001a79;" />

Comeback to

https://keras.io/examples/timeseries/timeseries_anomaly_detection/

<h3 style="color: #001a79;">Numenta Anomaly Benchmark (NAB) Dataset</h3>

<hr style="border-top: 1px solid #001a79;" />

- The Numenta Anomaly Benchmark (NAB) is a novel benchmark for evaluating algorithms for anomaly detection in streaming, online applications. 
- It is comprised of over 50 labeled real-world and artificial timeseries data files. 
- Plus a novel scoring mechanism designed for real-time applications.
- All of the data and code is fully open-source, with extensive documentation, and a scoreboard of anomaly detection algorithms found <a href="https://github.com/numenta/NAB" target="_blank">here</a>. 

Source: <a href="https://github.com/numenta/NAB" target="_blank">Kaggle: Numenta Anomaly Benchmark (NAB) Dataset</a>

<h3 style="color: #001a79;">Loading the Data</h3>

<hr style="border-top: 1px solid #001a79;" />

To demonstrate anomaly detection effectively we will use two datasets from the Numenta Anomaly Benchmark (NAB) Dataset. 

1. <a href="https://github.com/numenta/NAB/blob/master/data/artificialNoAnomaly/art_daily_small_noise.csv" target="_blank">`art_daily_small_noise.csv`</a> file for training.
2. <a href="https://github.com/numenta/NAB/blob/master/data/artificialWithAnomaly/art_daily_jumpsup.csv" target="_blank">`the art_daily_jumpsup.csv`</a> file for testing.

### `df_small_noise`

In [ ]:
# These CSVs can be found at the follow URL.
root_url = 'https://raw.githubusercontent.com/numenta/NAB/master/data/'

In [7]:
# art_daily_small_noise.csv data file url.
no_anomaly_url = root_url + 'artificialNoAnomaly/art_daily_small_noise.csv'

In [16]:
# Save data as a df - df_small_noise. 
# parse_dates=True will tell pandas you are importing a date while index_col="timestamp" sets timestamp as the index. 
parse_dates=True
df_small_noise = pd.read_csv(no_anomaly_url, parse_dates=True, index_col="timestamp")
df_small_noise

value
timestamp                     
2014-04-01 00:00:00  18.324919
2014-04-01 00:05:00  21.970327
2014-04-01 00:10:00  18.624806
2014-04-01 00:15:00  21.953684
2014-04-01 00:20:00  21.909120
...                        ...
2014-04-14 23:35:00  20.211195
2014-04-14 23:40:00  20.531041
2014-04-14 23:45:00  19.032127
2014-04-14 23:50:00  21.798801
2014-04-14 23:55:00  18.682620

[4032 rows x 1 columns]

`df_small_noise` contains 4,032 rows of data with 2 columns `timestamp` (the index) and `value`. 

`timestamp` - <br>
`value` - 

### The Daily Jumpsup Dataframe 

In [11]:
# Data file with anomaly.
with_anomaly_url = root_url + 'artificialWithAnomaly/art_daily_jumpsup.csv'

df_daily_jumpsup  = pd.read_csv(with_anomaly_url, parse_dates=True, index_col="timestamp")
df_daily_jumpsup .head()

value
timestamp                     
2014-04-01 00:00:00  19.761252
2014-04-01 00:05:00  20.500833
2014-04-01 00:10:00  19.961641
2014-04-01 00:15:00  21.490266
2014-04-01 00:20:00  20.187739